File: leetcode/307.区域和检索 - 数组可修改.py
class NumArray:
    def lowbit(self, x):
        return x & -x # 利用负数的补码
    def __init__(self, nums: List[int]):
        self.n = len(nums)
        self.nums = nums
        self.tr = [0] * (self.n + 1)
        for i, x in enumerate(nums):
            self.insert(i, x)
    def insert(self, i, val):
        i += 1
        while i <= self.n:
            self.tr[i] += val
            i += self.lowbit(i)
    def update(self, index: int, val: int) -> None:
        t = val - self.nums[index]
        self.nums[index] = val
        index += 1
        while index <= self.n:
            self.tr[index] += t
            index += self.lowbit(index)
    def query(self, idx):
        res = 0
        while idx > 0:
            res += self.tr[idx]
            idx -= self.lowbit(idx)
        return res
    def sumRange(self, left: int, right: int) -> int:
        return self.query(right + 1) - self.query(left)


# Your NumArray object will be instantiated and called as such:
# obj = NumArray(nums)
# obj.update(index,val)
# param_2 = obj.sumRange(left,right)

File: leetcode/307.区域和检索 - 数组可修改.cpp

const int N = 100010;
typedef long long LL;
int n;
int tr[N];
int a[N];

class NumArray {
public:

    int lowbit(int x){
        return x & -x;
    }
    int query(int x){
        LL res = 0;
        for(int i = x; i; i -= lowbit(i)){
            res += tr[i];
        }
        return res;
    }
    
    void update_tr(int x, int c){
        for(int i = x; i <= n; i += lowbit(i)){
            tr[i] += c;
        }
    }

    NumArray(vector<int>& nums) {
        //构造函数
        memset(tr, 0, sizeof tr);
        n = nums.size();
        for(int i = 0; i < n; i++){
            a[i+1] = nums[i];
        }
        for(int i = 0; i < n; i++){
            update_tr(i+1, nums[i]);
        }
    }
    
    void update(int index, int val) {
        update_tr(index+1, val-a[index+1]);
        a[index+1] = val;
    }
    
    int sumRange(int left, int right) {
        return query(right+1) - query(left);
    }
};

/**
 * Your NumArray object will be instantiated and called as such:
 * NumArray* obj = new NumArray(nums);
 * obj->update(index,val);
 * int param_2 = obj->sumRange(left,right);
 */